File: code/sorting/src/topological_sort/Topological.py
from collections import defaultdict 

class Graph: 
    def __init__(self,vertices): 
        self.graph = defaultdict(list) 
        self.V = vertices 
  
    
    def addEdge(self,u,v): 
        self.graph[u].append(v) 
  
    
    def topologicalSortUtil(self,v,visited,stack): 
  
        visited[v] = True
        for i in self.graph[v]: 
            if visited[i] == False: 
                self.topologicalSortUtil(i,visited,stack) 

        stack.insert(0,v) 

    def topologicalSort(self): 
       
        visited = [False]*self.V 
        stack =[] 
        for i in range(self.V): 
            if visited[i] == False: 
                self.topologicalSortUtil(i,visited,stack) 
  
        print(stack)
print("Enter the size of the directed graph")
n=int(input()) 
print("Enter the edges of the directed graph")
g= Graph(n) 
for i in range(0,n):
    val=str(input())
    arr=val.split(" ")
    edge1=int(arr[0])
    edge2=int(arr[1])
    g.addEdge(edge1,edge2)

print("Topological Sort of the given graph-:")
g.topologicalSort() 

File: code/Most_Useful_Algorithms/Topological.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <unordered_map>
#include <unordered_set>

using namespace std;

class Graph {
public:
    int V;
    unordered_map<int, vector<int>> adj;

    Graph(int vertices) : V(vertices) {}

    
    void addEdge(int u, int v) {
        adj[u].push_back(v);
    }


    void topologicalSortUtil(int v, unordered_set<int>& visited, stack<int>& s) {
        visited.insert(v);

        for (const int& neighbor : adj[v]) {
            if (visited.find(neighbor) == visited.end()) {
                topologicalSortUtil(neighbor, visited, s);
            }
        }

        s.push(v);
    }

    
    void topologicalSort() {
        unordered_set<int> visited;
        stack<int> s;

        for (const auto& pair : adj) {
            int v = pair.first;
            if (visited.find(v) == visited.end()) {
                topologicalSortUtil(v, visited, s);
            }
        }

        cout << "Topological Sorting Order:" << endl;
        while (!s.empty()) {
            cout << s.top() << " ";
            s.pop();
        }
        cout << endl;
    }
};

int main() {
    int V, E;
    cout << "Enter the number of vertices (V) and edges (E): ";
    cin >> V >> E;

    Graph g(V);

    cout << "Enter the directed edges (format: u v):" << endl;
    for (int i = 0; i < E; i++) {
        int u, v;
        cin >> u >> v;
        g.addEdge(u, v);
    }

    g.topologicalSort();

    return 0;
}
